src/sdi-notify.c

```diff
@@ -35,12 +36,126 @@ struct _SdiNotify {
+    if (names == NULL)
+      return NULL;
+    return g_strdup(names[0]);
+  }
+
+  if (G_IS_FILE_ICON(icon)) {
+    GFileIcon *file_icon = G_FILE_ICON(icon);
+    // the value returned by g_file_icon_get_file() is owned by the instance
+    return g_file_get_path(g_file_icon_get_file(file_icon));
+  }
+  return NULL;
+}
+
+typedef struct {
+  SdiNotify *self;
+  GVariant *snaps;
+} IgnoreNotifyData;
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(NotifyNotification, g_object_unref)
+
+static void app_close_notification(NotifyNotification *notification,
+                                   char *action, gpointer user_data) {
+  g_object_unref(notification);
+}
+
+static void app_ignore_snaps_notification(NotifyNotification *notification,
+                                          char *action, gpointer user_data) {
+  IgnoreNotifyData *data = user_data;
+  sdi_notify_action_ignore(NULL, data->snaps, data->self);
+  g_variant_unref(data->snaps);
+  g_free(user_data);
+
+  g_object_unref(notification);
+}
+
+static void show_pending_update_notification(SdiNotify *self,
+                                             const gchar *title,
+                                             const gchar *body, GIcon *icon,
+                                             GSList *snaps) {
+  g_autofree gchar *icon_name = get_icon_name_from_gicon(icon);
+  // Don't use g_autoptr because it must survive for the actions
+  NotifyNotification *notification =
+      notify_notification_new(title, body, icon_name);
+  if (icon_name != NULL) {
+    // don't use g_autoptr with the GVariant because it is consumed in set_hint
+    GVariant *hint_icon = g_variant_new_string(icon_name);
+    notify_notification_set_hint(notification, "image-path", hint_icon);
+  }
+  notify_notification_add_action(notification, "app.close-notification",
+                                 _("Close"), app_close_notification, NULL,
+                                 NULL);
+  notify_notification_add_action(notification, "default", _("Close"),
+                                 app_close_notification, NULL, NULL);
+  IgnoreNotifyData *data = g_malloc(sizeof(IgnoreNotifyData));
+  data->self = self;
+  data->snaps = get_snap_list(snaps);
+  notify_notification_add_action(notification, "app.ignore-notification",
+                                 _("Ignore"), app_ignore_snaps_notification,
+                                 data, NULL);
+  notify_notification_show(notification, NULL);
+}
+
+static void show_simple_notification(SdiNotify *self, const gchar *title,
+                                     const gchar *body, GIcon *icon,
+                                     const gchar *id) {
+  g_autofree gchar *icon_name = get_icon_name_from_gicon(icon);
+  // Don't use g_autoptr because it must survive for the actions
+  NotifyNotification *notification =
+      notify_notification_new(title, body, icon_name);
+
+  if (icon_name != NULL) {
+    // don't use g_autoptr with the GVariant because it is consumed in set_hint
```

Comment:
Could drop this comment by using one line: `notify_notification_set_hint(notification, "image-path", g_variant_new_string(icon_name));`

Reply:
Fixed.